### PR TITLE
add WeightedOceanMSE to criterion

### PR DESCRIPTION
**Pull request by zacespinosa**
# Modulus Pull Request

## Description

I added a new loss function to the criterion module that allows for channel weighting and LSM weighting. This allows us to train a multi-channel DLOM while applying the LSM.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/metrics/climate/healpix_loss.py [hunk 1]
```diff
@@ -189,3 +189,68 @@ def forward(self, prediction, target, average_channels=True):
...
+        self.lsm_sum = None
+        self.lsm_var_sum = None
+        self.loss_weights = th.tensor(weights)
+
+    def setup(self, trainer):
+        """
+        reshape lsm and put on device 
+        pushes weights to cuda device 
+        """
+        ### 1. OCEAN PREP ###
```
Comment (yairchn) [inline modulus/metrics/climate/healpix_loss.py:228]: in the recent training data we curated at Nvidia land sea mask is a channel in the zarr training data, this implies that a separate lsm file is found, is there a reason for this?

File: modulus/metrics/climate/healpix_loss.py [hunk 2]
```diff
@@ -189,3 +189,68 @@ def forward(self, prediction, target, average_channels=True):
...
+        self.lsm_ds = xr.open_dataset(self.lsm_file,**self.open_dict).constants.sel(self.selection_dict)
+        # 1-lsm gives the percentage of pixel that has ocean
```
Comment (yairchn) [inline modulus/metrics/climate/healpix_loss.py:230]: I am not sure I know the discussion leading to this, but you are using "percentage of pixel that has ocean" implying a weight that is not just 0 or 1 but could be in between. What informs this choice ? for the model resolution a data point is either full ocean or full atmosphere right ?

Comment (zacespinosa) [reply modulus/metrics/climate/healpix_loss.py:230]: The ERA5 lsm is 0s and 1s. When we interpolate to HPX we do not force the interpolated values to be strictly binary. This is because we want coastal cells (which include both ocean and land) to truthfully represent the fraction of land/ocean they contain.

File: modulus/metrics/climate/healpix_loss.py (hunk 1 above)
Comment (zacespinosa) [reply modulus/metrics/climate/healpix_loss.py:228]: I've implemented this to be consistent with the current implementation of OceanMSE (line 169 in this file) which also loads an external LSM file. Do you know how the DLOMs you or other NVIDIA folks have trained handled this?

I agree that eventually this should be updated for both the OceanMSE and WeightedOceanMSE implementations.

Comment (nathanielcresswellclay) [reply modulus/metrics/climate/healpix_loss.py:228]: We need to have access to the LSM within the loss class to provide weights to the loss. The way it's implemented here, we pass the path to a file which contains the LSM data.

This this file could/should be the training dataset in most cases.

File: modulus/metrics/climate/healpix_loss.py (hunk 2 above)
Comment (nathanielcresswellclay) [reply modulus/metrics/climate/healpix_loss.py:230]: There wasn't much discussion regarding how best to treat coastal pixels. The simplest solution-- using the the HEALPix mapped version of LSM (with values between 0 and 1)-- worked well so we kept it. I suggest keeping it simple unless someone can think of a compelling reason to change.

**Review by nathanielcresswellclay — APPROVED** [review]
Looks good! Definitely necessary for future models with expanded DLOMs . 

Not sure if you've added corresponding configs to NV-dlesm or not yet, but we should think about the best way to share configs.
